### PR TITLE
feat: [#441] Optimize the HTTP testing process [#1]

### PR DESCRIPTION
**Pull request by kkumar-gcc**
## 📑 Description
Usage Example:
```go
func (s *ExampleTestSuite) TestIndex() {
	response, err := s.Http(s.T()).WithHeader("X-Header", "value").Get("/users/1")
	s.Nil(err)
	response.AssertOk()
	response.AssertHeader("X-Header", "value")
}
```

RelatedTo https://github.com/goravel/goravel/issues/441

## Summary by CodeRabbit

- **New Features**
	- Introduced `TestRequest` and `TestResponse` interfaces for enhanced HTTP request and response testing.
	- Added methods for managing headers and cookies in HTTP requests.
	- Implemented comprehensive assertion methods for validating HTTP response statuses and headers.
	- Added a new method to the `TestCase` struct for simplified HTTP request testing.
	- Introduced a suite of unit tests for validating various HTTP response behaviors.

- **Bug Fixes**
	- Added a new error variable `RouteFacadeNotSet` to handle uninitialized route facades.

- **Chores**
	- Updated dependency versions in the `go.mod` file.

## ✅ Checks

- [ ] Added test cases for my code

### CODE REVIEW

**Comment by kkumar-gcc** [comment]
@coderabbitai review

File: go.mod [hunk 1]
```diff
@@ -4,16 +4,14 @@ go 1.22.0
 
 require (
 	github.com/RichardKnop/machinery/v2 v2.0.13
-	github.com/brianvoe/gofakeit/v7 v7.1.1
-	github.com/brianvoe/gofakeit/v7 v7.1.1
+	github.com/brianvoe/gofakeit/v6 v6.28.0
```
Comment (hwbrzzl) [inline go.mod:7]: Downgrade?

File: go.mod [hunk 2]
```diff
@@ -44,6 +42,8 @@ require (
 	gorm.io/plugin/dbresolver v1.5.3
 )
 
+require github.com/go-sql-driver/mysql v1.8.1 // indirect
```
Comment (hwbrzzl) [inline go.mod:45]: Move it to the below section.

File: testing/test_request.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+func NewTestRequest(t *testing.T) contractstesting.TestRequest {
+	return &TestRequest{
+		t:              t,
+		defaultHeaders: make(map[string]string),
+		defaultCookies: make(map[string]string),
+	}
+}
+
+func (m *TestRequest) WithHeaders(headers map[string]string) contractstesting.TestRequest {
```
Comment (hwbrzzl) [inline testing/test_request.go:27]: We can use `r` in all structs (consistent with other structs), except test suit (it's `s`).

```suggestion
func (r *TestRequest) WithHeaders(headers map[string]string) contractstesting.TestRequest {
```

File: testing/test_response.go [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+func (r *TestResponseImpl) AssertInternalServerError() contractstesting.TestResponse {
+	return r.AssertStatus(http.StatusInternalServerError)
+}
+
+func (r *TestResponseImpl) AssertServiceUnavailable() contractstesting.TestResponse {
+	return r.AssertStatus(http.StatusServiceUnavailable)
+}
+
+func (r *TestResponseImpl) AssertHeader(headerName, value string) contractstesting.TestResponse {
+	color.Errorln(r.Response.Header)
```
Comment (hwbrzzl) [inline testing/test_response.go:143]: Debug? Can be removed?

File: testing/test_response.go [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+func (r *TestResponseImpl) AssertCookieMissing(name string) contractstesting.TestResponse {
+	assert.Nil(r.t, r.getCookie(name), fmt.Sprintf("Cookie [%s] is present on response.", name))
+
+	return r
+}
+
+func (r *TestResponseImpl) T() *testing.T {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
```
Comment (hwbrzzl) [inline testing/test_response.go:194]: Is it necessary?

File: testing/test_response.go [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+	return r.t
+}
+
+func (r *TestResponseImpl) getStatusCode() int {
+	return r.Response.StatusCode
+}
+
+func (r *TestResponseImpl) getContent() (content string, err error) {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
```
Comment (hwbrzzl) [inline testing/test_response.go:204]: Ditto, the assertions should be run one by one, is there a chance of conflict?

File: testing/test_response.go [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+	if r.content != "" {
+		return r.content, err
+	}
+	if r.Response.Body == nil {
+		return "", fmt.Errorf("response body is nil")
+	}
+
+	defer func(Body io.ReadCloser) {
+		err = Body.Close()
+	}(r.Response.Body)
```
Comment (hwbrzzl) [inline testing/test_response.go:215]:
```suggestion
	defer r.Response.Body.Close()
```

**Review by hwbrzzl — COMMENTED** [review]
Pretty good 👍 and additional problems:

1. CI failed;
2. Solve AI's issues;

File: testing/test_response.go (hunk 6 above)
Comment (kkumar-gcc) [reply testing/test_response.go:204]: If someone runs it concurrently?

File: go.mod (hunk 1 above)
Comment (kkumar-gcc) [reply go.mod:7]: Must because of my local clone with upgrade it

File: testing/test_response.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+func (r *TestResponseImpl) AssertTooManyRequests() contractstesting.TestResponse {
+	return r.AssertStatus(http.StatusTooManyRequests)
+}
+
+func (r *TestResponseImpl) AssertInternalServerError() contractstesting.TestResponse {
+	return r.AssertStatus(http.StatusInternalServerError)
+}
+
```
[the remark this answers is not in the document]
Comment (kkumar-gcc) [reply testing/test_response.go:133]: Please write suggest unit tests

File: contracts/testing/test_request.go [hunk 9]
```diff
@@ -0,0 +1,10 @@
+package testing
+
+type TestRequest interface {
+	WithHeaders(headers map[string]string) TestRequest
+	WithHeader(key, value string) TestRequest
+	WithoutHeader(key string) TestRequest
+	WithCookies(cookies map[string]string) TestRequest
```
Comment (hwbrzzl) [inline contracts/testing/test_request.go:7]:
```suggestion
	WithCookies(cookies map[string]any) TestRequest
```

File: go.mod [hunk 10]
```diff
@@ -167,3 +165,5 @@ require (
 	modernc.org/memory v1.7.2 // indirect
 	modernc.org/sqlite v1.28.0 // indirect
 )
+
+require github.com/go-sql-driver/mysql v1.8.1 // indirect
```
Comment (hwbrzzl) [inline go.mod:169]:
```suggestion
	modernc.org/sqlite v1.28.0 // indirect
	require github.com/go-sql-driver/mysql v1.8.1 // indirect
)
```

File: testing/test_response.go [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+)
+
+type TestResponseImpl struct {
+	t        *testing.T
+	response *http.Response
+	content  string
+}
+
+func NewTestResponse(t *testing.T, response *http.Response) contractstesting.TestResponse {
+	//content := recorder.Body.String()
```
Comment (hwbrzzl) [inline testing/test_response.go:21]:
```suggestion
```

**Review by hwbrzzl — DISMISSED** [review]
Great PR 👍

File: contracts/testing/test_request.go (hunk 9 above)
Comment (kkumar-gcc) [reply contracts/testing/test_request.go:7]: I am going to cast it to string(http.Cookie expects value of type string) anyways so it's better to be done by developers

**Review by hwbrzzl — APPROVED** [review]

**Pull request merged**